File: packages/nats-client-ts/tests/connect-rejects-stale-wss.test.ts
/**
 * RED [unit] — a stale persisted non-`tls://` nats_url is rejected at
 * `connect()` BEFORE any transport dispatch (ts client).
 *
 * Two cases: the `wss://`-non-localhost case (verify-only) AND the
 * NEW `ws://localhost` / `wss://localhost` cases — RED today, since localhost
 * is still a bypass on current `main`.
 *
 * Scenario: `config.json` still carries a non-`tls://` nats_url — either a
 * `wss://<non-localhost>` (persisted before the cutover) or a stale
 * `ws://localhost` / `wss://localhost` (persisted while localhost was a
 * plaintext bypass, before it was removed). The host is upgraded to the
 * guard-collapsed client without re-registering. `doConnect` runs the shared
 * guard before the transport branch, so the stale url must throw synchronously
 * with NO `wsconnect` and NO node-TCP `connect` attempt (both mocked; assert
 * neither fired) — no hang against a dead ws:// / wss:// endpoint.
 *
 * QA-owned (adversarial-testing). NEVER weaken. Do NOT re-widen the guard to
 * accept `ws://localhost` so the localhost cases pass — the bypass is deleted.
 */

import { mkdtempSync, rmSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { afterEach, beforeEach, describe, expect, it, vi } from "vitest";

const { wsMock, tcpMock } = vi.hoisted(() => {
  const fakeNc = {
    isClosed: () => false,
    status: async function* () {
      /* no status events */
    },
    drain: async () => undefined,
  };
  return {
    wsMock: vi.fn().mockResolvedValue(fakeNc),
    tcpMock: vi.fn().mockResolvedValue(fakeNc),
  };
});

vi.mock("@nats-io/nats-core", async (importOriginal) => {
  const orig = await importOriginal<typeof import("@nats-io/nats-core")>();
  return {
    ...orig,
    wsconnect: wsMock,
    credsAuthenticator: vi.fn(() => vi.fn()),
  };
});
vi.mock("@nats-io/transport-node", () => ({ connect: tcpMock }));

const { KlodiClient } = await import("../src/client.js");

let home: string;

function makeClient(natsUrl: string): InstanceType<typeof KlodiClient> {
  home = mkdtempSync(join(tmpdir(), "klodi-stale-wss-"));
  const configPath = join(home, "config.json");
  const credsPath = join(home, "nats.creds");
  writeFileSync(
    configPath,
    JSON.stringify({
      handle: "alice",
      user_id: "u-1",
      nkey_public: "UAAAAAAAAAAAA",
      nats_url: natsUrl,
    }),
  );
  writeFileSync(credsPath, "-----BEGIN NATS USER JWT-----\nfake\n", {
    mode: 0o600,
  });
  return new KlodiClient({ credsPath, configPath });
}

beforeEach(() => {
  vi.clearAllMocks();
});

afterEach(() => {
  if (home) rmSync(home, { recursive: true, force: true });
});

describe("connect() rejects a stale persisted non-tls:// url before transport dispatch", () => {
  // `wss://<non-localhost>` was already rejected by the guard collapse
  // (verify-only); the localhost forms are the flip introduced here.
  it.each([
    "wss://klodi-net.4gpts.com",
    "ws://localhost:8080",
    "wss://localhost",
  ])(
    "throws on %s and dispatches neither transport (no wsconnect, no node-TCP, no hang)",
    async (staleUrl) => {
      const client = makeClient(staleUrl);
      await expect(client.connect()).rejects.toThrow();
      expect(wsMock).not.toHaveBeenCalled();
      expect(tcpMock).not.toHaveBeenCalled();
      expect(client.isConnected()).toBe(false);
    },
  );
});

File: packages/nats-client-rs/tests/connect_rejects_stale_wss_test.rs
//! RED [unit] — a stale persisted non-`tls://` nats_url is rejected at
//! `KlodiClient::connect()` BEFORE any transport dispatch (rs client).
//!
//! Two cases: the `wss://`-non-localhost case (verify-only) AND the
//! NEW `ws://localhost` / `wss://localhost` cases — RED today, since localhost
//! is still a bypass on current `main`.
//!
//! Scenario: `config.json` still carries a non-`tls://` nats_url — either a
//! `wss://<non-localhost>` (persisted before the cutover) or a stale
//! `ws://localhost` / `wss://localhost` (persisted while localhost was a
//! plaintext bypass, before it was removed). The host is upgraded to the
//! guard-collapsed client without re-registering. `connect()` runs the shared
//! guard (`client.rs:137`) BEFORE building `ConnectOptions` or entering
//! `initial_connect`, so the stale url must fail closed synchronously with a
//! structured `KlodiError::InvalidConfig` — no dial attempt.
//!
//! async-nats gives no transport-injection seam, so the no-dispatch proof is
//! terminality-under-timeout: `initial_connect` retries Io/Dns/TimedOut
//! FOREVER (`max_reconnects(None)`), so were the guard to run late, a dial
//! against the dead `wss://` endpoint would hang past the timeout. A prompt
//! `InvalidConfig` proves the guard fired ahead of dispatch. (Same
//! bounded-terminal recipe the TLS-trust suite uses.)
//!
//! Compiles against existing public symbols (`KlodiClient` / `KlodiError`),
//! so this binary is behavioural-RED (wss:// currently accepted → hang →
//! timeout), not compile-RED.
//!
//! QA-owned (adversarial-testing). NEVER weaken.

use std::path::{Path, PathBuf};
use std::time::{Duration, SystemTime, UNIX_EPOCH};

use klodi_nats_client::{KlodiClient, KlodiError};

/// Dependency-free unique temp dir (nats-client-rs has no `tempfile`
/// dev-dep — mirror the existing tests' `std::env::temp_dir` + nonce).
fn unique_dir(tag: &str) -> PathBuf {
    let nanos = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_nanos())
        .unwrap_or(0);
    let mut p = std::env::temp_dir();
    p.push(format!("klodi-stale-wss-{}-{}-{tag}", std::process::id(), nanos));
    std::fs::create_dir_all(&p).expect("create temp dir");
    p
}

fn write_session(dir: &Path, nats_url: &str) -> (PathBuf, PathBuf) {
    let config_path = dir.join("config.json");
    let creds_path = dir.join("nats.creds");
    std::fs::write(
        &config_path,
        serde_json::json!({
            "handle": "alice",
            "user_id": "u-1",
            "nkey_public": "UAAAAAAAAAAAA",
            "nats_url": nats_url,
        })
        .to_string(),
    )
    .expect("write config.json");
    std::fs::write(&creds_path, "-----BEGIN NATS USER JWT-----\nfake\n")
        .expect("write nats.creds");
    (creds_path, config_path)
}

#[tokio::test]
async fn connect_rejects_stale_wss_before_transport_dispatch() {
    let dir = unique_dir("home");
    let (creds_path, config_path) = write_session(&dir, "wss://klodi-net.4gpts.com");

    let client = KlodiClient::new(&creds_path, &config_path)
        .await
        .expect("new() only loads config+creds; it does not run the guard");

    // The timeout is the no-hang tripwire: a late guard would let
    // initial_connect retry the dead wss:// endpoint forever.
    let outcome = tokio::time::timeout(Duration::from_secs(5), client.connect()).await;

    let result = outcome.expect(
        "connect() must return promptly (guard before dispatch), not hang on a \
         wss:// transport dial",
    );
    let err = result.expect_err("stale wss:// nats_url must be rejected by the guard");
    assert!(
        matches!(err, KlodiError::InvalidConfig(_)),
        "rejection must be a structured InvalidConfig error, got: {err:?}"
    );
    assert!(
        !client.is_connected().await,
        "no half-open connection may remain after a rejected connect()"
    );

    let _ = std::fs::remove_dir_all(&dir);
}

#[tokio::test]
async fn connect_rejects_stale_ws_localhost_before_transport_dispatch() {
    // The transport-guard flip: a stale
    // `ws://localhost` / `wss://localhost` was accepted at connect while
    // localhost was a bypass. After the collapse the guard rejects it with
    // `InvalidConfig` BEFORE any `ConnectOptions` / dial — no hang against a
    // localhost endpoint that no longer speaks ws://.
    for stale_url in ["ws://localhost:8080", "wss://localhost"] {
        let dir = unique_dir("home");
        let (creds_path, config_path) = write_session(&dir, stale_url);

        let client = KlodiClient::new(&creds_path, &config_path)
            .await
            .expect("new() only loads config+creds; it does not run the guard");

        // The timeout is the no-hang tripwire: a late guard would let the
        // client dial the localhost endpoint (or spin the reconnect loop).
        let outcome = tokio::time::timeout(Duration::from_secs(5), client.connect()).await;

        let result = outcome.unwrap_or_else(|_| {
            panic!("connect() must return promptly (guard before dispatch) for {stale_url}, not hang")
        });
        let err = result.expect_err(&format!(
            "stale {stale_url} must be rejected by the collapsed guard"
        ));
        assert!(
            matches!(err, KlodiError::InvalidConfig(_)),
            "rejection must be structured InvalidConfig for {stale_url}, got: {err:?}"
        );
        assert!(
            !client.is_connected().await,
            "no half-open connection may remain after a rejected connect() for {stale_url}"
        );

        let _ = std::fs::remove_dir_all(&dir);
    }
}
